include/arpa/nameser.h

```diff
@@ -308,15 +309,44 @@ typedef enum __ns_cert_types {
 #define NS_PUT16(s, cp) ns_put16((s), ((cp)+=2)-2)
 #define NS_PUT32(l, cp) ns_put32((l), ((cp)+=4)-4)
 
-unsigned ns_get16(const unsigned char *);
-unsigned long ns_get32(const unsigned char *);
-void ns_put16(unsigned, unsigned char *);
-void ns_put32(unsigned long, unsigned char *);
-
-int ns_initparse(const unsigned char *, int, ns_msg *);
-int ns_parserr(ns_msg *, ns_sect, int, ns_rr *);
-int ns_skiprr(const unsigned char *, const unsigned char *, ns_sect, int);
-int ns_name_uncompress(const unsigned char *, const unsigned char *, const unsigned char *, char *, size_t);
+// ns_get16 returns as an unsigned the concatenation of the 2 bytes pointed by cp in reversed order.
+unsigned ns_get16(const unsigned char *cp : count(2));
+
+// ns_get32 returns as an unsigned long the concatenation of the 4 bytes pointed by cp in reversed order.
+unsigned long ns_get32(const unsigned char *cp : count(4));
+
+// ns_get16 puts l as 2 bytes in reverse to cp.
+void ns_put16(unsigned l, unsigned char *cp : count(2));
+
+// ns_get32 puts l as 4 bytes in reverse to cp.
+void ns_put32(unsigned long l, unsigned char *cp : count(4));
+
+// ns_initparse fills in the data structure pointed to by handle using the response message
+// buffer msg of size len.
+int ns_initparse(const unsigned char *msg : count(len) itype(_Array_ptr<const unsigned char>),
+	int len,
+	ns_msg *handle : itype(_Ptr<ns_msg>));
+
+// ns_parserr extracts information about a response record pointed by handle and stores it
+// in rr, which is a parameter passed to other name server libarary routines.
+int ns_parserr(ns_msg *handle : itype(_Ptr<ns_msg>),
+	ns_sect section,
+	int rrnum,
+	ns_rr *rr : itype(_Ptr<ns_rr>));
+
+int ns_skiprr(const unsigned char *cp : bounds(cp, eom),
+	const unsigned char *eom : itype(_Ptr<const unsigned char>),
+	ns_sect section,
+	int count);
+
+// ns_name_uncompress expands a "compressed" domain name in src within the message msg and stores
+// the result to dst which has size dst_size. eom is a pointer to the first byte after the message.
+// It is used to make sure that ns_name_uncompress doesn't go past the end of the message.
+int ns_name_uncompress(const unsigned char *msg : bounds(msg, eom),
+	const unsigned char *eom : itype(_Ptr<const unsigned char>),
+	const unsigned char *src : bounds(src, eom),
+	char *dst : count(dst_size - 1) itype(_Nt_array_ptr<char>),
```

Comment:
Why is the upper bounds for `dst` `dst_size - 1` instead of `dst_size`?

Reply:
I think it should be `dst_size`. Thanks for the catch!